cmd/root.go

```diff
@@ -0,0 +1,150 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/hierynomus/taipan"
+	home "github.com/mitchellh/go-homedir"
+	"github.com/rb3ckers/trafficmirror/internal/config"
+	"github.com/rb3ckers/trafficmirror/internal/proxy"
+	"github.com/rs/zerolog"
+	"github.com/spf13/cobra"
+)
+
+var (
+	Version string
+	Commit  string
+	Date    string
+)
+
+var EnvPrefix = "TRAFFICMIRROR_"
+
+func RootCommand(cfg *config.Config) *cobra.Command {
+	var verbosity int
+
+	cmd := &cobra.Command{
+		Use:   "trafficmirror",
+		Short: "Runs Traffic Mirror",
+		Long: `
+HTTP proxy that:
+* sends requests to a main endpoint from which the response is returned
+* can mirror the requests to any additional number of endpoints
+
+Additional targets are configured via PUT/DELETE on the '/targets?url=<endpoint>'.
+`,
+		Version: fmt.Sprintf("%s (Built on: %s, Commit: %s)", Version, Date, Commit),
+		PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
+			switch verbosity {
+			case 0:
+				// Nothing to do
+			case 1:
+				zerolog.SetGlobalLevel(zerolog.InfoLevel)
+			case 2: //nolint:gomnd
+				zerolog.SetGlobalLevel(zerolog.DebugLevel)
+			default:
+				zerolog.SetGlobalLevel(zerolog.TraceLevel)
+			}
+
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			PrintUsage(cfg)
+			if err := RunProxy(cfg); err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	cmd.PersistentFlags().CountVarP(&verbosity, "verbose", "v", "Print more verbose logging")
+
+	cmd.Flags().StringP("listen", "l", ":8080", "Address to listen on and mirror traffic from")
+	cmd.Flags().StringP("main", "m", "http://localhost:8888", "Main proxy target, its responses will be returned to the client")
+	cmd.Flags().String("targets", "targets", "Path on which additional targets to mirror to can be added/deleted/listed via PUT, DELETE and GET")
+	cmd.Flags().String("targets-address", "", "Address on which the targets endpoint is made available. Leave empty to expose it on the address that is being mirrored")
+	cmd.Flags().String("password", "", "Provide a file that contains username/password to protect the configuration 'targets' endpoint. Contains 1 username/password combination separated by ':'.")
+	cmd.Flags().Int("fail-after", 30, "Remove a target when it has been failing for this many minutes.") //nolint:gomnd
+	cmd.Flags().Int("retry-after", 1, "After 5 successive failures a target is temporarily disabled, it will be retried after this many minutes.")
+	cmd.Flags().StringSlice("mirror", []string{}, "Mirror traffic to additional targets")
```

Comment:
I prefer to rephrase this to something like "Start with mirroring to these targets"